cancel_token/token.py

```diff
@@ -95,21 +95,32 @@ def raise_if_triggered(self) -> None:
         for token in self._chain:
             futures.append(asyncio.ensure_future(token.wait(), loop=self.loop))
 
-        def cancel_not_done(fut: 'asyncio.Future[None]') -> None:
+        try:
+            done, pending = await asyncio.wait(
+                futures,
+                return_when=asyncio.FIRST_COMPLETED,
+                loop=self.loop,
+            )
+        except asyncio.CancelledError as err:
+            # Since we use return_when=asyncio.FIRST_COMPLETED above, we can
+            # be sure none of our futures will be done here, so we don't need
+            # to check if any is done before cancelling.
             for future in futures:
-                if not future.done():
-                    future.cancel()
-
-        async def _wait_for_first(futures: Sequence[Awaitable[Any]]) -> None:
-            for future in asyncio.as_completed(futures):
-                # We don't need to catch CancelledError here (and cancel not done futures)
-                # because our callback (above) takes care of that.
-                await cast(Awaitable[Any], future)
-                return
-
-        fut = asyncio.ensure_future(_wait_for_first(futures), loop=self.loop)
-        fut.add_done_callback(cancel_not_done)
-        await fut
+                future.cancel()
+            for future in futures:
+                try:
+                    await future
+                except asyncio.CancelledError:
+                    pass
+            raise err
+        else:
+            for future in pending:
+                future.cancel()
+            for future in pending:
+                try:
+                    await future
+                except asyncio.CancelledError:
```

Comment:
6. This one fixes the cancellation and awaiting of the upstream cancel tokens.

Reply:
This is now covered in `test_token_wait_finishing_cleans_up_chained_token_waits`